day3/main.go

```diff
@@ -0,0 +1,266 @@
+
+	for i, line := range input {
+		for j, char := range line {
+			var count int
+			var mul int64
+			if char == '*' {
+
+				// has a digit before in the same line
+				if j != 0 && isDigit(line[j-1]) {
+					count++
+
+					for _, localNum := range numsInInput[i] {
+						if localNum.start <= j-1 && localNum.end >= j-1 {
+							if mul == 0 {
+								mul = localNum.num
+							} else {
+								mul *= localNum.num
+							}
+							break
+						}
+					}
+				}
+
+				// has a digit after in the same line
+				if j != len(line)-1 && isDigit(line[j+1]) {
+					count++
+
+					for _, localNum := range numsInInput[i] {
+						if localNum.start <= j+1 && localNum.end >= j+1 {
+
+							if mul == 0 {
+								mul = localNum.num
+							} else {
+								mul *= localNum.num
+							}
+							break
+						}
+					}
+				}
+
+				// has number/s in the line above or below
+				start := 0
+				if j != 0 {
+					start = j - 1
+				}
+				end := j
+				if j < len(line)-1 {
+					end += 1
+				}
+
+				// if line above has digit/s
+				if i != 0 {
+
+					for _, localNum := range numsInInput[i-1] {
+						// this is the tricky bit. Nums can exist in three different cases:
+						// note: '(' denotes beginning digit of number and ')' denotes the last digit
+						// note: '|' denote the start and beginning of the range where digit will be considered adjacent.
+						// 1. ----(-------)---- ie a single num spans the entire 3 'adjacent' indices
+						//    ------|-*-|------
+						// 2. ----(---)-------- ie a num end in one of the allowed indices.
+						//    ------|-*-|------
+						// 3. ---------(----)-- ie a num begins in one of the allowed indices.
+						//    ------|-*-|------
+						// considering cases 2 & 3, we can have 2 numbers adjacent to a '*' in the lines above or below, eg:
+						// ---123-456----
+						// ------*-------
+
+						// case 1
+						if (localNum.start <= start && localNum.end >= end) ||
+							// case 2
+							(localNum.start <= start && localNum.end <= end && localNum.end >= start) ||
+							// case 3
+							(localNum.start >= start && localNum.start <= end && localNum.end >= end) {
+							count++
+
+							if mul == 0 {
+								mul = localNum.num
+							} else {
+								mul *= localNum.num
+							}
+
+						}
+					}
+
+				}
+
+				// if line below has digit/s
+				if i != len(input)-1 {
+
+					// same logic as above applied here, this time for lines below '*'
+					for _, localNum := range numsInInput[i+1] {
+						if (localNum.start <= start && localNum.end >= end) || (localNum.start <= start && localNum.end <= end && localNum.end >= start) || (localNum.start >= start && localNum.start <= end && localNum.end >= end) {
+							count++
+							if mul == 0 {
+								mul = localNum.num
+							} else {
+								mul *= localNum.num
+							}
+
+						}
+					}
+				}
+
+			}
+
+			if count == 2 {
+				sum += mul
+			}
+		}
+	}
+	return sum
+}
+
+func isDigit(char byte) bool {
```

Comment:
Hi @ShraddhaAg

I noticed that you implemented a custom `isDigit` function instead of using the built-in [unicode.IsDigit()](https://pkg.go.dev/unicode#IsDigit). Is there a reason for this?

My solution uses a sliding window, which, in part 1, looks for any symbols surrounding a digit. In part 2, the window is fixed to size 4 (since the length of part numbers is always <= 3) and is anchored to a `*`, then it searches for part numbers in that window. Although there's a bit of repeated work, this approach helped me keep the logic relatively simple.

I'm also curious to see a regex solution! Thanks for all the contributions so far.